modules/new_serverpod_auth/serverpod_auth_email_account/serverpod_auth_email_account_server/lib/src/business/email_accounts.dart

```diff
@@ -0,0 +1,314 @@
+import 'dart:math';
+
+import 'package:serverpod/serverpod.dart';
+import 'package:serverpod_auth_email_account_server/serverpod_auth_email_account_server.dart';
+import 'package:serverpod_auth_email_account_server/src/business/password_hash.dart';
+import 'package:serverpod_auth_email_account_server/src/generated/protocol.dart';
+import 'package:serverpod_auth_email_account_server/src/util/random_string.dart';
+
+/// Email account management functions.
+abstract final class EmailAccounts {
+  /// Returns the [AuthUser]'s ID upon successful login.
+  ///
+  /// Throws [EmailAccountLoginException] for expected error cases.
+  static Future<UuidValue> login(
+    final Session session, {
+    required String email,
+    required String password,
+  }) async {
+    email = email.trim().toLowerCase();
+    password = password.trim();
+
+    var account = await EmailAccount.db.findFirstRow(
+      session,
+      where: (final t) => t.email.equals(email),
+    );
+
+    account ??=
+        await _importExistingUser(session, email: email, password: password);
+
+    if (account == null) {
+      throw EmailAccountLoginException(
+        reason: EmailAccountLoginFailureReason.invalidCredentials,
+      );
+    }
+
+    if (await _hasTooManyFailedSignIns(session, email)) {
+      throw EmailAccountLoginException(
+        reason: EmailAccountLoginFailureReason.tooManyAttempts,
+      );
+    }
+
+    if (!PasswordHash.validateHash(
```

Comment:
Nit: maybe we should move this into the query, so we never have the user around if the authentication was not checked.

But then that would only work for the current hash function, if we ever need to switch this in the future (as we did in the past), we'd have to check which one is employed and compute the hash correctly.